### ISSUE TITLE
tf.estimator.BoostedTreesEstimator returns all zero predictions when using custom loss_fn involving linear loss (zero second derivative)

### ISSUE THREAD

**Issue by valkenburg** [tensorflow/tensorflow#48119]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): **No**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **macos Big Sur && Google Colab**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: **na**
- TensorFlow installed from (source or binary): **pip (python 3.8)**
- TensorFlow version (use command below): **v2.4.0-49-g85c8b2a817f 2.4.1**
- Python version: **3.8**
- Bazel version (if compiling from source): **na**
- GCC/Compiler version (if compiling from source): **na**
- CUDA/cuDNN version: **na**
- GPU model and memory: **na**

**Describe the current behavior**
When creating a BoostedTreesEstimator with a custom head in order to use various loss functions, the predictions are always zero if the loss function involves tf.math.abs (I have only tried tf.math.abs and tf.math.pow, tf.math.pow works as expected). 

**Describe the expected behavior**
non-zero predictions. See the example and use custom_loss_mse instead of custom_loss_mae. In this example of uniform data in [0,1], both mse and mae should return predictions close to 0.5. 

**Standalone code to reproduce the issue**
[Colab gist](https://colab.research.google.com/gist/valkenburg/e35ce97d89523413a3db378174db9a6b/bootsedzeros.ipynb)
```python
import tensorflow as tf
import numpy as np

LABEL_DIM = 1
INPUT_DIMS = [1]

features = [
    tf.feature_column.numeric_column(f"random_input_{i}",
                                     shape = (sh,),
                                     dtype=tf.float32
                                    ) 
    for i, sh in enumerate(INPUT_DIMS)
]

def input_fn(num_samples = 1000):
    inputs = {f.key : np.random.rand(num_samples, s) for f, s in zip(features, INPUT_DIMS)}
    targets = np.random.rand(num_samples, LABEL_DIM)
    ds = tf.data.Dataset.from_tensor_slices((inputs, targets))
    ds = ds.shuffle(num_samples).repeat(10).batch(10) 
    return ds

def less_input_fn():
    return input_fn(10)

def custom_loss_mse(labels, logits):
    return tf.math.pow(labels - logits, 2)

def custom_loss_mae(labels, logits):
    return tf.math.abs(labels - logits)

est = tf.estimator.BoostedTreesEstimator(
    feature_columns = features,
    head = tf.estimator.RegressionHead(
        label_dimension = LABEL_DIM, 
#         loss_fn = custom_loss_mse,
        loss_fn = custom_loss_mae,
    ),
    n_batches_per_layer = 1,
    center_bias = True,
)

est.train(input_fn)

list(est.predict(less_input_fn))
```

*I posted [this question](https://stackoverflow.com/questions/66831261/tf-estimator-boostedtreesestimator-returns-all-zero-predictions-when-using-custo) in stackoverflow, just in case the behaviour is expected and I'm misunderstanding things.*

**Comment by amahendrakar** [comment]
@rmothukuru,
I was able to reproduce the issue with TF v2.4, TF v2.5.0rc0 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/55466ea74b2fc652dced96fbde778ca4/48119.ipynb). Thanks!

**Comment by valkenburg** [comment]
Reading up on this, https://blog.tensorflow.org/2019/03/how-to-train-boosted-trees-models-in-tensorflow.html mentions "Users can also choose to use any twice differentiable custom loss (by providing it to BoostedTreesEstimator)."

I think this is the culprit. The loss in my gist is twice differentiable most of the time, albeit that the second derivative is always zero, except at the origin `labels - logits == 0`.

So I think the issue could either be resolved by clearer documentation of the API (explaining the exact conditions on the loss), or a runtime detection of this invalid use with a non-cryptic exception.

However, I do not have the level of understanding about the implementation to say if the use in the gist is actually valid or invalid, and if there still is a bug to be fixed.

@wvzytlyn

**Comment by sushreebarsa** [comment]
@valkenburg Estimators are deprecated.Could you please refer to this [link ](https://www.tensorflow.org/api_docs/python/tf/estimator/BoostedTreesEstimator)  and try to follow the [migration](https://www.tensorflow.org/guide/migrate) document for more details ?Thanks!
